src/Core/Resolvers/CosmosMutationEngine.cs

```diff
@@ -23,17 +25,20 @@ namespace Azure.DataApiBuilder.Core.Resolvers
     public class CosmosMutationEngine : IMutationEngine
     {
         private readonly CosmosClientProvider _clientProvider;
-        private readonly ISqlMetadataProvider _metadataProvider;
+        private readonly IMetadataProviderFactory _metadataProviderFactory;
         private readonly IAuthorizationResolver _authorizationResolver;
+        private readonly RuntimeConfigProvider _runtimeConfigProvider;
 
         public CosmosMutationEngine(
             CosmosClientProvider clientProvider,
-            ISqlMetadataProvider metadataProvider,
-            IAuthorizationResolver authorizationResolver)
+            IMetadataProviderFactory metadataProviderFactory,
```

Comment:
I did a little more digging. Asp.net core DI container allows you to inject multiple implementations of an interface when your constructor denotes a parameter like `IEnumerable<IMetadataProvider> metadataProviders`. see https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-6.0#service-registration-methods

Would that be a potential mechanism to avoid the additional factory pattern abstraction added? Then the class itself (e.g. CosmosMutationEngine) would then be responsible for parsing through that list and either saving the relevant metadata providers, such as use case of multiple cosmos sources added via config. 

This may allow you to avoid the factory pattern and directly state your dependencies in the constructor.

Reply:
Hey Sean, so this is the approach i had taken earlier. But we had discussed the case of static web apps for which above approach would not work. I would need to register the metadataproviders per db by going through the runtimeconfig, so that in the factory the Ienumerable is populated with them. However, In some cases I cannot access the runtimeConfig immediately in startup and need to wait till its available. So we decided to go with this factory pattern as this is the same precedence and format used by the cosmos client provider as well.

Reply:
Created issue for this based on discussion offline: https://github.com/Azure/data-api-builder/issues/1754